---
title: "1.1-Combine phyloseq objects"
author: "Siyuan Ma"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
    df_print: paged
---
# Overview
- Combine studies into one phyloseq object.
- Filter to only CD/UC/control samples with available sample type.
- Add metadata columns (baseline indicator, read depth, etc.)

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir = "../")
```

```{r setup2, include=FALSE}
smar::sourceDir("functions/", recursive = FALSE)
setup("./")
```

# Combine studies into one phyloseq object

```{r load datasets and aggregate studies}
print(studies)
l_otus <- l_species <- l_genera <- list()
for(study in studies) {
  load(paste0("data/physeq/",
              study,
              "/otus.RData"))
  load(paste0("data/physeq/",
              study,
              "/species.RData"))
  load(paste0("data/physeq/",
              study,
              "/genera.RData"))
  l_otus[[study]] <- physeq
  l_species[[study]] <- physeq_species
  l_genera[[study]] <- physeq_genera
}
physeq_otus_prefilter <- combine_phyloseq(l_otus)
physeq_species_prefilter <- combine_phyloseq(l_species)
physeq_genera_prefilter <- combine_phyloseq(l_genera)
df_metadata <- smar::sample_data2(physeq_species_prefilter)
```

# Filter by disease and sample type

```{r filtering by sample type}
df_metadata <- df_metadata %>% 
  dplyr::filter(disease %in% c("CD", "UC", "control")) %>% 
  dplyr::filter(!is.na(sample_type))
```

# Add columns to metadata

```{r add read depth}
tb_readCounts <- studies %>% 
  purrr::map_dfr(~ paste0(dir_processed, 
                          "processed/", 
                          .x, 
                          "/16s/all_samples_read_counts.tsv") %>% 
                   readr::read_tsv(col_types = "cddd"))
df_metadata <- df_metadata %>% 
  dplyr::left_join(tb_readCounts, by = c("sample_accession_16S" = "# sample")) %>% 
  dplyr::rename(original_read_count = `original read count`,
                reads_mapping_to_OTU_with_taxonomy = `reads mapping to OTU with taxonomy`,
                reads_mapping_to_unclassified_OTU = `reads mapping to unclassifed OTU`)
# sanity check
# df_metadata %>% 
#   dplyr::filter(is.na(`original read count`),
#                 is.na(`reads mapping to OTU with taxonomy`),
#                 is.na(`reads mapping to unclassifed OTU`)) %>% 
#   nrow()
df_metadata$`read_depth` <- 
  apply(smar::otu_table2(physeq_otus_prefilter), 2, sum)[df_metadata$sample_accession_16S]
# sanity check
ncol(df_metadata) == ncol(smar::sample_data2(physeq_otus_prefilter)) + 4
```

```{r add baseline indicator}
# baseline indicator is the first observation in longitudinal studies
df_metadata_HMP2 <- df_metadata %>% 
  dplyr::filter(dataset_name == "HMP2") %>% 
  dplyr::mutate(time_point2 = as.numeric(time_point)) %>% 
  dplyr::group_by(subject_accession) %>% 
  dplyr::arrange(time_point2) %>% 
  dplyr::mutate(is_baseline = time_point2 %in% time_point2[1]) %>% 
  dplyr::select(-time_point2)
df_metadata_PROTECT <- df_metadata %>% 
  dplyr::filter(dataset_name == "PROTECT") %>% 
  dplyr::mutate(time_point2 = as.numeric(time_point)) %>% 
  dplyr::group_by(subject_accession) %>% 
  dplyr::arrange(time_point2) %>% 
  dplyr::mutate(is_baseline = time_point2 %in% time_point2[1]) %>% 
  dplyr::select(-time_point2)
df_metadata_LSS_PRISM <- df_metadata %>% 
  dplyr::filter(dataset_name == "LSS-PRISM") %>% 
  dplyr::mutate(time_point2 = as.numeric(time_point)) %>% 
  dplyr::group_by(subject_accession) %>% 
  dplyr::arrange(time_point2) %>% 
  dplyr::mutate(is_baseline = time_point2 %in% time_point2[1]) %>% 
  dplyr::select(-time_point2)
df_metadata_Jansson <- df_metadata %>% 
  dplyr::filter(dataset_name == "Jansson_Lamendella_Crohns") %>% 
  dplyr::mutate(time_point2 = time_point %>% 
                  stringr::str_replace_all("\\:.*$", "") %>% 
                  as.numeric) %>% 
  dplyr::group_by(subject_accession) %>% 
  dplyr::arrange(time_point2) %>% 
  dplyr::mutate(is_baseline = time_point2 %in% time_point2[1]) %>% 
  dplyr::select(-time_point2)
df_metadata_Herfarth <- df_metadata %>% 
  dplyr::filter(dataset_name == "Herfarth_CCFA_Microbiome_3B_combined") %>% 
  dplyr::mutate(time_point2 = as.Date(time_point, format = "%Y-%m-%d")) %>% 
  dplyr::group_by(subject_accession) %>% 
  dplyr::arrange(time_point2) %>% 
  dplyr::mutate(is_baseline = time_point2 %in% time_point2[1]) %>% 
  dplyr::select(-time_point2)
# for not longitudinal studies, baseline indicator is NA
df_metadata <- rbind(df_metadata_HMP2,
                     df_metadata_PROTECT,
                     df_metadata_LSS_PRISM,
                     df_metadata_Jansson,
                     df_metadata_Herfarth) %>% 
  dplyr::right_join(df_metadata, by = colnames(df_metadata))
# sanity check
ncol(df_metadata) == ncol(smar::sample_data2(physeq_otus_prefilter)) + 5
all(df_metadata$sample_accession_16S %in% sample_names(physeq_otus_prefilter))
class(df_metadata) <- "data.frame"
rownames(df_metadata) <- df_metadata$sample_accession_16S
```

# Prune features that are just not there after subsetting and save the combined phyloseq objects

```{r prune features save}
physeq_otus_prefilter <- phyloseq::prune_samples(rownames(df_metadata), 
                                                physeq_otus_prefilter) %>% 
  phyloseq::filter_taxa(function(x) any(x > 0), prune = TRUE)
sample_data(physeq_otus_prefilter) <- df_metadata
physeq_species_prefilter <- phyloseq::prune_samples(rownames(df_metadata), 
                                                   physeq_species_prefilter) %>% 
  phyloseq::filter_taxa(function(x) any(x > 0), prune = TRUE)
sample_data(physeq_species_prefilter) <- df_metadata
physeq_genera_prefilter <- phyloseq::prune_samples(rownames(df_metadata), 
                                                  physeq_genera_prefilter) %>% 
  phyloseq::filter_taxa(function(x) any(x > 0), prune = TRUE)
sample_data(physeq_genera_prefilter) <- df_metadata

save(physeq_otus_prefilter, file = "data/physeq/otus_prefilter.RData")
save(physeq_species_prefilter, file = "data/physeq/species_prefilter.RData")
save(physeq_genera_prefilter, file = "data/physeq/genera_prefilter.RData")
```