---
title: "2-QC"
author: "Siyuan Ma"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
    df_print: paged
---
# Overview
- Aggregate studies into one phyloseq object.
- Sensitivity check on mapping otus at a particular taxonamy rank.
- Taxa distribution across studies (alpha diveristy).
- Uniquely present/absent taxa.
- Library size distribution.
- Filtering.
```{r setup, echo=FALSE}
knitr::opts_knit$set(root.dir = normalizePath(".."))
```
```{r setup2, echo=FALSE}
rm(list = ls())
for(i.function in list.files("functions/", full.names = TRUE)) {
  source(i.function)
}
setup(getwd())
dir_output <- "results/2-QC/"
dir.create(dir_output, recursive = TRUE, showWarnings = FALSE)
```
# Aggregate studies
```{r load datasets and aggregate studies}
print(studies)
l_otus <- l_species <- l_genera <- list()
for(study in studies) {
  load(paste0("data/phyloseq/",
              study,
              "/otus.RData"))
  load(paste0("data/phyloseq/",
              study,
              "/species.RData"))
  load(paste0("data/phyloseq/",
              study,
              "/genera.RData"))
  l_otus[[study]] <- phylo
  l_species[[study]] <- phylo_species
  l_genera[[study]] <- phylo_genera
}
phylo_otus <- combine_phyloseq(l_otus)
phylo_species <- combine_phyloseq(l_species)
phylo_genera <- combine_phyloseq(l_genera)
```
# Grouping otus into genera seems more appropriate than species
```{r sensitivity taxonamy}
tb_long <- l_otus %>% 
  purrr::map_dfr(
    function(phylo_otus) {
      tb_ra_otus <- phylo_otus %>% 
        transform_sample_counts(tss) %>% 
        phyloseq_to_tb()
      tb_otu_taxonamy <- phylo_otus %>% 
        tax_table2() %>% 
        tibble::as_tibble(rownames = "feature") %>% 
        dplyr::mutate(classification = dplyr::case_when(
          Rank7 != "s__" ~ "fully classified",
          Rank1 == "k__" ~ "unclassified at kingdom",
          Rank2 == "p__" ~ "unclassified at phylumn",
          Rank3 == "c__" ~ "unclassified at class",
          Rank4 == "o__" ~ "unclassified at order",
          Rank5 == "f__" ~ "unclassified at family",
          Rank6 == "g__" ~ "unclassified at genus",
          Rank7 == "s__" ~ "unclassified at species"
        ) %>% 
          factor(levels= c("fully classified",
                           "unclassified at species",
                           "unclassified at genus",
                           "unclassified at family",
                           "unclassified at order",
                           "unclassified at class",
                           "unclassified at phylumn",
                           "unclassified at kingdom")))
      
      tb_ra_otus %>% 
        dplyr::left_join(tb_otu_taxonamy,
                         by = c("feature", "Rank1", "Rank2", "Rank3", 
                                "Rank4", "Rank5", "Rank6", "Rank7"))
    }
  )
p <- tb_long %>% 
  dplyr::group_by(dataset_name, feature, classification) %>%
  dplyr::filter(!duplicated(feature)) %>% dplyr::summarise(n = n()) %>%
  ggplot(aes(x = dataset_name, y = n, fill = classification)) +
  geom_bar(stat = "identity") +
  rotate_xaxis(30) +
  ggtitle("OTUs are mostly classified at order, family, genus, and species level")
print(p) %>% 
  ggsave(paste0(dir_output, "fig_OTUClassification.pdf"),
         .,
         width = 8,
         height = 5)

p <- tb_long %>% 
  dplyr::group_by(dataset_name, rownames, classification) %>% 
  dplyr::summarise(abundance = sum(abundance)) %>% dplyr::ungroup() %>% 
  dplyr::group_by(dataset_name, rownames) %>% 
  dplyr::mutate(abundance_fullyClassified = 
                  sum(abundance[classification %in% c("fully classified",
                                                      "unclassified at species")])) %>% dplyr::ungroup() %>% 
  dplyr::arrange(dataset_name, abundance_fullyClassified) %>% dplyr::mutate(rownames = factor(rownames, levels = unique(rownames))) %>% 
  ggplot(aes(x = rownames, y = abundance, fill = classification)) +
  geom_bar(stat = "identity") +
  facet_grid(.~dataset_name, scales = "free_x", space = "free_x") +
  no_label_xaxis() +
  ggtitle("OTUs are most abundantly classified at family, genus, and species level.")
print(p) %>% 
  ggsave(paste0(dir_output, "fig_speciesClassificationAbd.pdf"),
         .,
         width = 20,
         height = 6)

tb_long_speciesUnclassified <- tb_long %>% 
  dplyr::filter(classification == "unclassified at species") %>% 
  dplyr::group_by(dataset_name, rownames, Rank6, Rank7) %>% dplyr::summarise(abundance = sum(abundance)) %>% dplyr::ungroup() %>% 
  dplyr::group_by(Rank6, Rank7) %>% dplyr::mutate(mean_abundance = mean(abundance)) %>% dplyr::ungroup() %>% 
  dplyr::mutate(taxonamy = ifelse(mean_abundance >= sort(unique(mean_abundance), decreasing = TRUE)[10],
                                  paste(Rank6, Rank7, sep = "|"), "others")) %>% 
  dplyr::arrange(dplyr::desc(mean_abundance)) %>% dplyr::mutate(taxonamy = factor(taxonamy, levels = unique(taxonamy))) %>%
  dplyr::group_by(dataset_name, rownames) %>% dplyr::mutate(all_abundance = sum(abundance)) %>% dplyr::ungroup() %>% 
  dplyr::arrange(dataset_name, all_abundance) %>% dplyr::mutate(rownames = factor(rownames, levels = unique(rownames)))
p1 <- tb_long_speciesUnclassified %>% 
  ggplot(aes(x = rownames, y = abundance, fill = taxonamy)) +
  geom_bar(stat = "identity") +
  facet_grid(.~dataset_name, scales = "free_x", space = "free_x") +
  no_label_xaxis() +
  ggtitle("Abundance of OTUs unclassified at species level")
p2 <- tb_long %>% 
  dplyr::filter(classification == "unclassified at species") %>% 
  dplyr::group_by(dataset_name) %>% 
  dplyr::filter(!duplicated(feature)) %>% 
  dplyr::ungroup() %>% 
  dplyr::mutate(taxonamy = paste(Rank6, Rank7, sep = "|")) %>% 
  dplyr::filter(taxonamy %in% levels(tb_long_speciesUnclassified$taxonamy)) %>% 
  dplyr::mutate(taxonamy = factor(taxonamy, levels = levels(tb_long_speciesUnclassified$taxonamy))) %>% 
  dplyr::group_by(dataset_name, taxonamy) %>% 
  dplyr::summarise(n_otus = n()) %>% 
  ggplot(aes(x = dataset_name, y = n_otus, fill = taxonamy)) +
  geom_bar(stat = "identity") +
  rotate_xaxis(30)
cowplot::plot_grid(p1, p2, nrow = 1, rel_widths = c(0.8, 0.2)) %>% 
  ggsave(paste0(dir_output, "fig_speciesUnclassified.pdf"),
         .,
         width = 25,
         height = 6)

tb_long_generaUnclassified <- tb_long %>% 
  dplyr::filter(classification == "unclassified at genus") %>% 
  dplyr::group_by(dataset_name, rownames, Rank5, Rank6) %>% dplyr::summarise(abundance = sum(abundance)) %>% dplyr::ungroup() %>% 
  dplyr::group_by(Rank5, Rank6) %>% dplyr::mutate(mean_abundance = mean(abundance)) %>% dplyr::ungroup() %>% 
  dplyr::mutate(taxonamy = ifelse(mean_abundance >= sort(unique(mean_abundance), decreasing = TRUE)[10],
                                  paste(Rank5, Rank6, sep = "|"), "others")) %>% 
  dplyr::arrange(dplyr::desc(mean_abundance)) %>% dplyr::mutate(taxonamy = factor(taxonamy, levels = unique(taxonamy))) %>%
  dplyr::group_by(dataset_name, rownames) %>% dplyr::mutate(all_abundance = sum(abundance)) %>% dplyr::ungroup() %>% 
  dplyr::arrange(dataset_name, all_abundance) %>% dplyr::mutate(rownames = factor(rownames, levels = unique(rownames)))
p1 <- tb_long_generaUnclassified %>% 
  ggplot(aes(x = rownames, y = abundance, fill = taxonamy)) +
  geom_bar(stat = "identity") +
  facet_grid(.~dataset_name, scales = "free_x", space = "free_x") +
  no_label_xaxis() +
  ggtitle("Abundance of OTUs unclassified at genera level")
p2 <- tb_long %>% 
  dplyr::filter(classification == "unclassified at genus") %>% 
  dplyr::group_by(dataset_name) %>% 
  dplyr::filter(!duplicated(feature)) %>% 
  dplyr::ungroup() %>% 
  dplyr::mutate(taxonamy = paste(Rank5, Rank6, sep = "|")) %>% 
  dplyr::filter(taxonamy %in% levels(tb_long_generaUnclassified$taxonamy)) %>% 
  dplyr::mutate(taxonamy = factor(taxonamy, levels = levels(tb_long_generaUnclassified$taxonamy))) %>% 
  dplyr::group_by(dataset_name, taxonamy) %>% 
  dplyr::summarise(n_otus = n()) %>% 
  ggplot(aes(x = dataset_name, y = n_otus, fill = taxonamy)) +
  geom_bar(stat = "identity") +
  rotate_xaxis(30)
p3 <- tb_long %>% 
  dplyr::filter(classification == "unclassified at genus",
                Rank5 == "f__Enterobacteriaceae",
                Rank6 == "g__") %>% 
  dplyr::group_by(dataset_name) %>% 
  dplyr::filter(!duplicated(feature)) %>% 
  dplyr::ungroup() %>% 
  dplyr::mutate(taxonamy = paste(Rank5, Rank6, sep = "|")) %>% 
  ggplot(aes(x = dataset_name)) +
  geom_bar() +
  rotate_xaxis(30) +
  ggtitle("# unclassified f_Enterobacteriacea OTUs")
cowplot::plot_grid(p1, p2, p3, nrow = 1, rel_widths = c(4, 1, 1)) %>% 
  ggsave(paste0(dir_output, "fig_generaUnclassified.pdf"),
         .,
         width = 30,
         height = 6)
```
# Examine alpha diversity.
```{r "alpha diversity across studies"}
tb_nTaxaSampleLibSize <- list(otus = l_otus,
                              species = l_species,
                              genera = l_genera) %>% 
  purrr::imap_dfr(function(l_phylo, feature_level) {
    purrr::imap_dfr(l_phylo, function(phylo, study) {
      tibble::tibble(
        feature_level = feature_level,
        study = study,
        N_taxa = phyloseq::ntaxa(phylo),
        N_samples = phyloseq::nsamples(phylo),
        median_log10LibSize = phylo %>% phyloseq::sample_sums() %>% 
          add(1) %>% log(10) %>% 
          median()
      )
    })
  })
p1 <- tb_nTaxaSampleLibSize %>% 
  ggplot(aes(x = N_samples,
             y = N_taxa)) +
  geom_point() +
  ggrepel::geom_text_repel(aes(label = study)) +
  facet_wrap(~factor(feature_level, levels = c("otus", "species", "genera")),
             scales = "free_y")
p2 <- tb_nTaxaSampleLibSize %>% 
  ggplot(aes(x = median_log10LibSize,
             y = N_taxa)) +
  geom_point() +
  ggrepel::geom_text_repel(aes(label = study)) +
  facet_wrap(~factor(feature_level, levels = c("otus", "species", "genera")),
             scales = "free_y")
cowplot::plot_grid(p1, p2, ncol = 1) %>% 
  cowplot_title("Alpha diversity across studies") %>% 
  ggsave(file = paste0(dir_output, "fig_alphaDiversity.pdf"),
         width = 15,
         heigh = 10)
```
# Examine uniquely present and absent taxa.
```{r "dataset specific species"}
# function to mark observations as outliers
is_outlier <- function(x) {
  return(x < quantile(x, 0.25) - 1.5 * IQR(x) | x > quantile(x, 0.75) + 1.5 * IQR(x))
}
# simplified study names
tb_1 <- readr::read_csv("tables/table1/assets/table1_metadata.csv")
all_studies <- setdiff(unique(tb_1$Study), NA_character_)
list(species = phylo_species,
     genera = phylo_genera) %>% 
  purrr::iwalk(function(phylo, feature_level) {
    tb_long <- phylo %>% 
      phyloseq::transform_sample_counts(tss) %>% 
      phyloseq_to_tb()
    tb_summarise <- tb_long %>% 
      dplyr::group_by(feature) %>% 
      dplyr::mutate(prevalence = mean(abundance > 1e-5)) %>% 
      dplyr::ungroup() %>% 
      dplyr::group_by(feature, dataset_name, prevalence) %>% 
      dplyr::summarise(prevalence_study = mean(abundance > 1e-5),
                       mean_abundance = mean(abundance)) %>% 
      dplyr::group_by(feature) %>% 
      dplyr::mutate(n_absent = sum(prevalence_study == 0),
                    max_mean_abundance = max(mean_abundance)) %>% 
      dplyr::ungroup() %>% 
      dplyr::left_join(tb_1, by = c("dataset_name" = "study_full_name")) %>% 
      dplyr::mutate(feature = feature %>% 
                      stringr::str_replace_all(stringr::fixed("|NA"), ""))
    p1 <- tb_summarise %>% 
      dplyr::filter(prevalence > 0.7) %>% 
      dplyr::group_by(feature) %>% 
      dplyr::mutate(spread = IQR(mean_abundance),
                    label = ifelse(is_outlier(mean_abundance),
                                   Study, 
                                   NA_character_),
                    mean_abundance_not_outlier = ifelse(is_outlier(mean_abundance),
                                                        NA_real_,
                                                        mean_abundance)) %>% 
      dplyr::ungroup() %>% 
      dplyr::arrange(spread) %>% 
      dplyr::mutate(feature = factor(feature, levels = unique(feature))) %>% 
      ggplot(aes(x = feature, y = mean_abundance)) +
      geom_boxplot() +
      geom_point(aes(x = feature, y = mean_abundance_not_outlier),
                 position = position_jitter(width = 0.2)) +
      ggrepel::geom_label_repel(aes(label = label),
                                angle = 270,
                                size = 2,
                                min.segment.length = 0) +
      xlab("Feature") + ylab("Study mean relative abundance") +
      coord_flip()
    ggsave(filename = paste0(dir_output, feature_level,
                             "highly_prevalent.pdf"),
           p1,
           width = 10,
           height = 8)
    p2 <- tb_summarise %>% 
      dplyr::filter(prevalence <= 0.7, n_absent == 0) %>% 
      dplyr::group_by(feature) %>% 
      dplyr::mutate(spread = IQR(prevalence_study),
                    label = ifelse(is_outlier(prevalence_study),
                                   dataset_name, 
                                   NA_character_),
                    prevalence_study_not_outlier = ifelse(is_outlier(prevalence_study),
                                                          NA_real_,
                                                          prevalence_study)) %>% 
      dplyr::ungroup() %>% 
      dplyr::arrange(spread) %>% 
      dplyr::mutate(feature = factor(feature, levels = unique(feature))) %>% 
      ggplot(aes(x = feature, y = prevalence_study)) +
      geom_boxplot() +
      geom_point(aes(y = prevalence_study_not_outlier), 
                 position = position_jitter(width = 0.2)) +
      ggrepel::geom_label_repel(aes(label = label),
                                angle = 270,
                                size = 2,
                                min.segment.length = 0) +
      xlab("Feature") + ylab("Study prevalence") +
      coord_flip()
    ggsave(filename = paste0(dir_output, feature_level,
                             "medium_prevalent.pdf"),
           p2,
           width = 10,
           height = 15)
    p3 <- tb_summarise %>% 
      dplyr::filter(prevalence <= 0.7, n_absent >= 1, prevalence_study > 0) %>% 
      dplyr::group_by(feature, n_absent) %>% 
      dplyr::mutate(n_prevalent = n(),
                    include = max(prevalence_study) > 0.05,
                    spread = IQR(prevalence_study),
                    label = ifelse(is_outlier(prevalence_study),
                                   dataset_name, 
                                   NA_character_),
                    prevalence_study_not_outlier = ifelse(is_outlier(prevalence_study),
                                                          NA_real_,
                                                          prevalence_study),
                    study_present = ifelse(n_absent[1] == 9,
                                           Study[1],
                                           NA_character_)) %>% 
      dplyr::ungroup() %>% 
      dplyr::filter(include) %>% 
      dplyr::arrange(-n_prevalent) %>% 
      dplyr::mutate(n_prevalent_group = paste0("#Studies prevalent ", n_prevalent) %>% 
                      forcats::as_factor()) %>% 
      dplyr::arrange(n_prevalent, study_present, spread, prevalence_study) %>% 
      dplyr::mutate(feature = factor(feature, levels = unique(feature))) %>% 
      ggplot(aes(x = feature, y = prevalence_study)) +
      geom_boxplot() +
      geom_point(aes(x = feature, y = prevalence_study_not_outlier),
                 position = position_jitter(width = 0.2)) +
      ggrepel::geom_label_repel(aes(label = label),
                                angle = 270,
                                size = 2,
                                min.segment.length = 0) +
      geom_text(aes(label = study_present,
                    y = ifelse(prevalence_study > 0.5,
                               prevalence_study - 0.05,
                               prevalence_study + 0.05), 
                    hjust = ifelse(prevalence_study > 0.5,
                                   1,
                                   0)), size = 2) +
      facet_wrap(.~n_prevalent_group, scales = "free_y", nrow = 2) +
      theme(axis.text.y = element_text(size = 5)) +
      xlab("Feature") + ylab("Study prevalence") +
      coord_flip()
    # p3 <- tb_summarise %>% 
    #   dplyr::filter(prevalence <= 0.7, n_absent >= 1, prevalence_study > 0) %>% 
    #   dplyr::group_by(feature, n_absent) %>% 
    #   dplyr::summarise(n_prevalent = n(),
    #                    include = max(prevalence_study) > 0.05,
    #                    study_present = ifelse(n_absent[1] == 9,
    #                                           Study[1],
    #                                           NA_character_)) %>% 
    #   dplyr::ungroup() %>% 
    #   dplyr::filter(include) %>% 
    #   dplyr::arrange(n_prevalent, study_missing, study_present) %>% 
    #   dplyr::mutate(feature = factor(feature, levels = unique(feature))) %>% 
    #   ggplot(aes(x = feature, y = n_prevalent)) +
    #   geom_bar(stat = "identity") +
    #   geom_text(aes(label = study_missing, y = n_prevalent + 0.3), hjust = 0, size = 2) +
    #   geom_text(aes(label = study_present, y = n_prevalent + 0.3), hjust = 0, size = 2) +
    #   scale_y_continuous(breaks = 1:9,
    #                      limits = c(0, 18)) +
    #   coord_flip()
    ggsave(filename = paste0(dir_output, feature_level,
                             "absent_at_least_one.pdf"),
           p3,
           width = 40,
           height = 23,
           limitsize = FALSE)
    
    # tb_presence <- tb_long %>% 
    #   dplyr::group_by(feature, dataset_name) %>% 
    #   dplyr::summarise(present = any(abundance > 0)) %>% 
    #   dplyr::group_by(feature) %>% 
    #   dplyr::mutate(n_present = sum(present),
    #                 n_absent = sum(!present))
    # tb_long <- tb_long %>% 
    #   dplyr::left_join(tb_presence,
    #                    by = c("feature", "dataset_name"))
    # tb_uniqPres <- tb_long %>% 
    #   dplyr::filter(n_present == 1, present) %>% 
    #   dplyr::group_by(feature, dataset_name) %>% 
    #   dplyr::summarise(mean_abundance = mean(abundance)) %>% 
    #   dplyr::ungroup()
    # tb_uniqAbs <- tb_long %>% 
    #   dplyr::filter(n_absent == 1, present) %>% 
    #   dplyr::group_by(feature) %>% 
    #   dplyr::summarise(mean_abundance = mean(abundance)) %>% 
    #   dplyr::ungroup() %>% 
    #   dplyr::left_join(
    #     tb_presence %>% 
    #       dplyr::filter(n_absent == 1,
    #                     !present),
    #     by = "feature"
    #   ) %>% 
    #   dplyr::select(feature, dataset_name, mean_abundance)
    # 
    # # Table summarises feature presence
    # tb_presence %>% 
    #   dplyr::filter(!duplicated(feature)) %>% 
    #   dplyr::group_by(n_present) %>% 
    #   dplyr::summarise(N = n()) %>% 
    #   t() %>% 
    #   DT::datatable(caption = feature_level) %>% 
    #   print()
    # 
    # # Figure uniquely present features
    # p <- tb_uniqPres %>% 
    #   dplyr::arrange(dataset_name, mean_abundance) %>% 
    #   dplyr::mutate(feature = factor(feature, levels = feature)) %>% 
    #   ggplot(aes(x = feature, y = mean_abundance, color = dataset_name)) + geom_point() +
    #   rotate_xaxis(90) 
    # ggsave(filename = paste0(dir_output, "fig_", feature_level, "UniquePresent.pdf"),
    #        plot = p,
    #        width = 40,
    #        height = 10)
    # p <- tb_uniqPres %>% 
    #   dplyr::filter(mean_abundance > 1e-4) %>% 
    #   dplyr::arrange(dataset_name, mean_abundance) %>% 
    #   dplyr::mutate(feature = factor(feature, levels = feature)) %>% 
    #   ggplot(aes(x = feature, y = mean_abundance, color = dataset_name)) + geom_point() +
    #   rotate_xaxis(90) + coord_flip()
    # p %>% print() %>% 
    #   ggsave(filename = paste0(dir_output, "fig_", feature_level, "UniquePresentZoom.pdf"),
    #          plot = .,
    #          width = 15,
    #          height = 10)
    # 
    # # Figure uniquely absent features
    # p <- tb_uniqAbs %>% 
    #   dplyr::arrange(dataset_name, mean_abundance) %>% 
    #   dplyr::mutate(feature = factor(feature, levels = feature)) %>% 
    #   ggplot(aes(x = feature, y = mean_abundance, color = dataset_name)) + geom_point() +
    #   rotate_xaxis(90) + coord_flip() 
    # p %>% print() %>% 
    #   ggsave(filename = paste0(dir_output, "fig_", feature_level, "UniqueAbsent.pdf"),
    #          plot = .,
    #          width = 15,
    #          height = 10)
  })
```
# Individual OTUS
```{r individual OTUS}
mat_tax_tmp <- tax_table2(l_species$Jansson_Lamendella_Crohns)
mat_tax_tmp %>% 
  tibble::as_tibble(rownames = "otu_cluster") %>% 
  dplyr::filter(Rank7 == "s__oncorhynchi")
mat_tax_tmp <- tax_table2(l_species$MucosalIBD)
mat_tax_tmp %>% 
  tibble::as_tibble(rownames = "otu_cluster") %>% 
  dplyr::filter(Rank7 == "s__blattae")
tax_table2(phylo_genera) %>% 
  tibble::as_tibble(rownames = "feature") %>% 
  dplyr::filter(Rank6 == "g__Escherichia")
```
# Examine library sizes
```{r check lib sizes, message=FALSE}
# sanity check
tb_1 <- readr::read_csv("tables/table1/assets/table1_metadata.csv")

tb_libSize <- list(otu = l_otus,
                   species = l_species, 
                   genus = l_genera) %>% 
  purrr::imap_dfr(function(l_phylo, feature_level) {
    l_phylo %>% purrr::imap_dfr(function(phylo, dataset_name) {
      tibble::tibble(
        sample_accession_16S = phyloseq::sample_names(phylo),
        dataset_name = dataset_name,
        lib_size = phyloseq::sample_sums(phylo),
        feature_level = feature_level)
    })
  })
# sanity check
# tb_libSize %>% 
#   dplyr::select(dataset_name, sample_accession_16S, feature_level, lib_size) %>% 
#   tidyr::spread(key = feature_level, value = lib_size) %>% 
#   ggplot(aes(x = genus, y = otu)) + geom_point() + ggtitle("lib size otus vs. genera level")
tb_libSize <- tb_libSize %>% dplyr::filter(feature_level == "species")
tb_readCounts <- studies %>% 
  purrr::map_dfr(~ paste0(dir_processed, 
                          "processed/", 
                          .x, 
                          "/16s/all_samples_read_counts.tsv") %>% 
                   readr::read_tsv(col_types = "cddd") %>% 
                   dplyr::mutate(dataset_name = .x)) %>% 
  dplyr::rename(original_read_count = `original read count`,
                reads_mapping_to_OTU_with_taxonomy = `reads mapping to OTU with taxonomy`,
                reads_mapping_to_unclassifed_OTU = `reads mapping to unclassifed OTU`)
tb_libSize <- tb_libSize %>% 
  dplyr::left_join(tb_readCounts,
                   c("dataset_name" = "dataset_name",
                     "sample_accession_16S" = "# sample"))
# sanity check
# the two are equal
# tb_libSize %>% 
#   ggplot(aes(x = lib_size, y = reads_mapping_to_OTU_with_taxonomy)) + geom_point() +
#   facet_wrap(~dataset_name) + ggtitle("The two should be the same")
tb_libSize_summary <- tb_libSize %>% 
  dplyr::left_join(tb_1, by = c("dataset_name" = "study_full_name")) %>% 
  dplyr::group_by(Study) %>% 
  dplyr::summarise(mean_libSize = mean(lib_size),
                   sd_libSize = sd(lib_size),
                   mean_original_libSize = mean(original_read_count),
                   sd_original_libSize = sd(original_read_count),
                   n = n(),
                   n_greater_than_3000 = sum(lib_size > 3000))
tb_libSize_summary %>% 
  dplyr::mutate(Study,
                `Total read count` = paste0(round(mean_original_libSize/1000)*1000,
                                            " (", round(sd_original_libSize/1000)*1000,
                                            ")"),
                `# reads with taxonomy mapping` = 
                  paste0(round(mean_libSize/1000)*1000,
                         " (", round(sd_libSize/1000)*1000,
                         ")"),
                `n (n with greater than 3,000 library size)` = 
                  paste0(n, " (", n_greater_than_3000, ")")) %>% 
  dplyr::select(Study, `Total read count`, `# reads with taxonomy mapping`, `n (n with greater than 3,000 library size)`) %>% 
  readr::write_tsv(paste0(dir_output, "library_size.tsv"))
tb_libSize %>% 
  tidyr::gather(key = "Type of read count",
                value = "N",
                original_read_count,
                lib_size,
                reads_mapping_to_OTU_with_taxonomy,
                reads_mapping_to_unclassifed_OTU,
                factor_key = TRUE) %>% 
  ggplot(aes(x = dataset_name, y = log10(N + 1), color = `Type of read count`)) + 
  geom_boxplot() + rotate_xaxis(90)
ggsave(filename = paste0(dir_output, "fig_libSize.pdf"),
       width = 10,
       height = 8)
# p1 <- tb_libSize %>% 
#   ggplot(aes(x = dataset_name, fill = (lib_size > 3000))) +
#   geom_bar(stat = "count") +
#   rotate_xaxis(30) +
#   ggtitle("Cutoff = 3000")
# p2 <- tb_libSize %>% 
#   ggplot(aes(x = dataset_name, fill = (lib_size > 1000))) +
#   geom_bar(stat = "count") +
#   rotate_xaxis(30) +
#   ggtitle("Cutoff = 1000")
# cowplot::plot_grid(p1, p2, ncol = 1) %>% 
#   cowplot_title(title = "Lib size filtering cut off?") %>% print() %>%
#   ggsave(filename = paste0(dir_output, "fig_libSizeCutoff.pdf"),
#          width = 10,
#          height = 8)
# sample_data(phylo_species) <- 
#   sample_data2(phylo_species) %>% 
#   tibble::rownames_to_column() %>% 
#   dplyr::left_join(tb_readCounts, 
#                    by = c("dataset_name" = "dataset_name", 
#                           "sample_accession_16S" = "# sample")) %>% 
#   tibble::column_to_rownames()
# sample_data(phylo_genera) <- 
#   sample_data2(phylo_genera) %>%
#   tibble::rownames_to_column() %>% 
#   dplyr::left_join(tb_readCounts, 
#                    by = c("dataset_name" = "dataset_name", 
#                           "sample_accession_16S" = "# sample")) %>% 
#   tibble::column_to_rownames()
```
# Filter samples and taxa
```{r Filter samples and taxa}
# sanity check
# libSize_genera <- phyloseq::sample_sums(phylo_genera)
# libSize_species <- phyloseq::sample_sums(phylo_species)
# all(libSize_species == libSize_genera)
# record library size before filtering
libSize <- phyloseq::sample_sums(phylo_genera)
sample_data(phylo_species) <- 
  sample_data(phylo_genera) <- 
  cbind(sample_data2(phylo_genera), 
        libSize)

phylo_species <- phylo_species %>% 
  phyloseq::subset_samples(libSize > 3000)
phylo_genera <- phylo_genera %>% 
  phyloseq::subset_samples(libSize > 3000)

cutoff_abundance <- c(1e-5, 1e-4, 1e-3)
cutoff_prevalence <- c(0.05, 0.1)
tb_param <-  tidyr::crossing(cutoff_abundance, cutoff_prevalence)
tb_filter <- (1:nrow(tb_param)) %>% 
  purrr::map_dfr(function(i_param) {
    cutoff_abundance <- tb_param$cutoff_abundance[i_param]
    cutoff_prevalence <- tb_param$cutoff_prevalence[i_param]
    list(phylo_genera,
         phylo_species) %>% 
      purrr::map2_dfr(c("genera", "species"), 
                      function(phylo, feature_level) {
                        phylo_rel <- phylo %>% 
                          phyloseq::transform_sample_counts(MMUPHin:::tss)
                        tb_presence <- phylo_rel %>% 
                          phyloseq_to_tb() %>% 
                          dplyr::group_by(feature, dataset_name) %>% 
                          dplyr::summarise(
                            n = n(),
                            present = sum(abundance > cutoff_abundance) > 
                              n*cutoff_prevalence
                          ) %>% 
                          dplyr::ungroup()
                        tb_presence %>% 
                          dplyr::group_by(feature) %>% 
                          dplyr::summarise(n_study = sum(present)) %>% 
                          dplyr::mutate(cutoff_abundance = cutoff_abundance,
                                        cutoff_prevalence = cutoff_prevalence,
                                        feature_level = feature_level)
                      })
  })
# tb_filter %>%
#   readr::write_tsv(path = paste0(dir_output, "filter.tsv"))
tb_filter_summary <- tb_filter %>%
  dplyr::group_by(feature_level, cutoff_abundance, cutoff_prevalence) %>%
  dplyr::summarise(nfeature = n(),
                   nfeature_nonUnique = sum(n_study > 1),
                   nfeature_unique = sum(n_study == 1)) %>%
  dplyr::arrange(feature_level,
                 dplyr::desc(cutoff_abundance), dplyr::desc(cutoff_prevalence)) %>%
  readr::write_tsv(path = paste0(dir_output, "filter_summarise.tsv")) 


phylo_genera <- 
  tb_filter %>% 
  dplyr::filter(feature_level == "genera",
                cutoff_abundance == 1e-4,
                cutoff_prevalence == 0.05,
                n_study >= 1) %>% 
  dplyr::pull(feature) %>% 
  phyloseq::prune_taxa(phylo_genera) %>% 
  prune_taxaSamples()
phylo_species <- 
  tb_filter %>% 
  dplyr::filter(feature_level == "species",
                cutoff_abundance == 1e-4,
                cutoff_prevalence == 0.05,
                n_study >= 1) %>% 
  dplyr::pull(feature) %>% 
  phyloseq::prune_taxa(phylo_species) %>% 
  prune_taxaSamples()
save(phylo_genera, file = "data/phyloseq/genera.RData")
save(phylo_species, file = "data/phyloseq/species.RData")
# save(phylo_otus, file = "data/phyloseq/otus.RData")
```

# Examine alpha diversity post filtering.
```{r "alpha diversity across studies"}
tb_1 <- readr::read_csv("tables/table1/assets/table1_metadata.csv")
tb_nTaxaSampleLibSize <- list(species = phylo_species,
                              genera = phylo_genera) %>% 
  purrr::imap_dfr(function(phylo, feature_level) {
    phylo %>% phyloseq_to_tb %>% 
      dplyr::group_by(dataset_name, feature) %>% 
      dplyr::summarise(present = any(abundance > 0),
                       median_lib_size = median(log10(libSize))) %>% 
      dplyr::group_by(dataset_name) %>% 
      dplyr::summarise(alpha = sum(present),
                       median_lib_size = unique(median_lib_size)) %>% 
      dplyr::ungroup() %>% 
      dplyr::mutate(feature_level = feature_level)
  })
tb_nTaxaSampleLibSize <- tb_nTaxaSampleLibSize %>% 
  dplyr::left_join(tb_1, by = c("dataset_name" = "study_full_name"))
p <- tb_nTaxaSampleLibSize %>% 
  dplyr::mutate(feature_level = feature_level %>% 
                  dplyr::recode_factor("species" = "Species",
                                       "genera" = "Genera")) %>% 
  ggplot(aes(x = median_lib_size,
             y = alpha)) +
  geom_point() +
  ggrepel::geom_text_repel(aes(label = Study)) +
  facet_wrap(~feature_level,
             scales = "free_y") +
  xlab("Median log10 library size") + ylab("#Present features")
ggsave(file = paste0(dir_output, "fig_alphaDiversity_post_filtering.pdf"),
       width = 8,
       heigh = 4)
```
# Examine uniquely present and absent taxa.
```{r "dataset specific species"}
# function to mark observations as outliers
is_outlier <- function(x) {
  return(x < quantile(x, 0.25) - 1.5 * IQR(x) | x > quantile(x, 0.75) + 1.5 * IQR(x))
}
# simplified study names
tb_1 <- readr::read_csv("tables/table1/assets/table1_metadata.csv")
all_studies <- setdiff(unique(tb_1$Study), NA_character_)
list(species = phylo_species) %>% 
  purrr::iwalk(function(phylo, feature_level) {
    tb_long <- phylo %>% 
      phyloseq::transform_sample_counts(tss) %>% 
      phyloseq_to_tb()
    tb_summarise <- tb_long %>% 
      dplyr::group_by(feature) %>% 
      dplyr::mutate(prevalence = mean(abundance > 1e-5)) %>% 
      dplyr::ungroup() %>% 
      dplyr::group_by(feature, dataset_name, prevalence) %>% 
      dplyr::summarise(prevalence_study = mean(abundance > 1e-5),
                       mean_abundance = mean(abundance)) %>% 
      dplyr::group_by(feature) %>% 
      dplyr::mutate(n_absent = sum(prevalence_study == 0),
                    max_mean_abundance = max(mean_abundance)) %>% 
      dplyr::ungroup() %>% 
      dplyr::left_join(tb_1, by = c("dataset_name" = "study_full_name")) %>% 
      dplyr::mutate(feature = feature %>% 
                      stringr::str_replace_all(stringr::fixed("|NA"), ""))
    p1 <- tb_summarise %>% 
      dplyr::filter(prevalence > 0.7) %>% 
      dplyr::group_by(feature) %>% 
      dplyr::mutate(spread = IQR(mean_abundance),
                    label = ifelse(is_outlier(mean_abundance),
                                   Study, 
                                   NA_character_),
                    mean_abundance_not_outlier = ifelse(is_outlier(mean_abundance),
                                                        NA_real_,
                                                        mean_abundance)) %>% 
      dplyr::ungroup() %>% 
      dplyr::arrange(spread) %>% 
      dplyr::mutate(feature = factor(feature, levels = unique(feature))) %>% 
      ggplot(aes(x = feature, y = mean_abundance)) +
      geom_boxplot(outlier.size = 0.5, outlier.alpha = 0.5) +
      geom_point(aes(x = feature, y = mean_abundance_not_outlier),
                 position = position_jitter(width = 0.2),
                 size = 0.5,
                 alpha = 0.5) +
      ggrepel::geom_text_repel(aes(label = label),
                               angle = 270,
                               size = 2,
                               min.segment.length = 0) +
      xlab("Feature") + ylab("Study mean relative abundance") +
      coord_flip()
    ggsave(filename = paste0(dir_output, feature_level,
                             "highly_prevalent_post_filtering.pdf"),
           p1,
           width = 12,
           height = 6)
    p2 <- tb_summarise %>% 
      dplyr::filter(prevalence <= 0.7, n_absent == 0) %>% 
      dplyr::group_by(feature) %>% 
      dplyr::mutate(spread = IQR(prevalence_study),
                    label = ifelse(is_outlier(prevalence_study),
                                   Study, 
                                   NA_character_),
                    prevalence_study_not_outlier = ifelse(is_outlier(prevalence_study),
                                                          NA_real_,
                                                          prevalence_study)) %>% 
      dplyr::ungroup() %>% 
      dplyr::arrange(spread) %>% 
      dplyr::mutate(feature = factor(feature, levels = unique(feature))) %>% 
      ggplot(aes(x = feature, y = prevalence_study)) +
      geom_boxplot(outlier.size = 0.5,
                   outlier.alpha = 0.5) +
      geom_point(aes(y = prevalence_study_not_outlier), 
                 position = position_jitter(width = 0.2),
                 size = 0.5,
                 alpha = 0.5) +
      ggrepel::geom_text_repel(aes(label = label),
                               angle = 270,
                               size = 2,
                               min.segment.length = 0) +
      xlab("Feature") + ylab("Study prevalence") +
      coord_flip()
    ggsave(filename = paste0(dir_output, feature_level,
                             "medium_prevalent_post_filtering.pdf"),
           p2,
           width = 12,
           height = 10)
    
    test <- tb_summarise %>% 
      dplyr::filter(prevalence <= 0.7, n_absent == 0) %>% 
      dplyr::left_join(tb_libSize_summary, by = "Study") %>% 
      dplyr::group_by(feature) %>% 
      dplyr::mutate(cor = cor(prevalence_study, 
                              mean_libSize,
                              method = "spearman"))
    p_test <- test %>% dplyr::group_by(feature) %>% 
      dplyr::mutate(spread = IQR(prevalence_study)) %>% 
      dplyr::ungroup() %>% 
      dplyr::filter(!duplicated(feature)) %>% 
      dplyr::arrange(spread) %>% 
      dplyr::mutate(feature = factor(feature, levels = unique(feature))) %>%
      ggplot(aes(x = feature, y = cor)) +
      geom_bar(stat = "identity") +
      coord_flip()
     ggsave(filename = paste0(dir_output, feature_level,
                             "medium_prevalent_post_filtering_cor.pdf"),
           p_test,
           width = 12,
           height = 10)
     
    p3 <- tb_summarise %>% 
      dplyr::filter(prevalence <= 0.7, n_absent >= 1, prevalence_study > 0) %>% 
      dplyr::group_by(feature, n_absent) %>% 
      dplyr::mutate(n_prevalent = n(),
                    include = max(prevalence_study) > 0.05,
                    spread = IQR(prevalence_study),
                    label = ifelse(is_outlier(prevalence_study),
                                   Study, 
                                   NA_character_),
                    prevalence_study_not_outlier = ifelse(is_outlier(prevalence_study),
                                                          NA_real_,
                                                          prevalence_study),
                    study_present = ifelse(n_absent[1] == 9,
                                           Study[1],
                                           NA_character_)) %>% 
      dplyr::ungroup() %>% 
      dplyr::filter(include) %>% 
      dplyr::arrange(-n_prevalent) %>% 
      dplyr::mutate(n_prevalent_group = paste0("#Studies prevalent ", n_prevalent) %>% 
                      forcats::as_factor()) %>% 
      dplyr::arrange(n_prevalent, study_present, spread, prevalence_study) %>% 
      dplyr::mutate(feature = factor(feature, levels = unique(feature))) %>% 
      ggplot(aes(x = feature, y = prevalence_study)) +
      geom_boxplot(outlier.size = 0.5,
                   outlier.alpha = 0.5) +
      geom_point(aes(x = feature, y = prevalence_study_not_outlier),
                 position = position_jitter(width = 0.2),
                 size = 0.5,
                 alpha = 0.5) +
      ggrepel::geom_text_repel(aes(label = label),
                               angle = 270,
                               size = 2,
                               min.segment.length = 0) +
      geom_text(aes(label = study_present,
                    y = ifelse(prevalence_study > 0.5,
                               prevalence_study - 0.05,
                               prevalence_study + 0.05), 
                    hjust = ifelse(prevalence_study > 0.5,
                                   1,
                                   0)), size = 2) +
      facet_wrap(.~n_prevalent_group, scales = "free_y", nrow = 3) +
      theme(axis.text.y = element_text(size = 5)) +
      xlab("Feature") + ylab("Study prevalence") +
      coord_flip()
    # p3 <- tb_summarise %>% 
    #   dplyr::filter(prevalence <= 0.7, n_absent >= 1, prevalence_study > 0) %>% 
    #   dplyr::group_by(feature, n_absent) %>% 
    #   dplyr::summarise(n_prevalent = n(),
    #                    include = max(prevalence_study) > 0.05,
    #                    study_present = ifelse(n_absent[1] == 9,
    #                                           Study[1],
    #                                           NA_character_)) %>% 
    #   dplyr::ungroup() %>% 
    #   dplyr::filter(include) %>% 
    #   dplyr::arrange(n_prevalent, study_missing, study_present) %>% 
    #   dplyr::mutate(feature = factor(feature, levels = unique(feature))) %>% 
    #   ggplot(aes(x = feature, y = n_prevalent)) +
    #   geom_bar(stat = "identity") +
    #   geom_text(aes(label = study_missing, y = n_prevalent + 0.3), hjust = 0, size = 2) +
    #   geom_text(aes(label = study_present, y = n_prevalent + 0.3), hjust = 0, size = 2) +
    #   scale_y_continuous(breaks = 1:9,
    #                      limits = c(0, 18)) +
    #   coord_flip()
    ggsave(filename = paste0(dir_output, feature_level,
                             "absent_at_least_one_post_filtering.pdf"),
           p3,
           width = 21,
           height = 15,
           limitsize = FALSE)
    
    
    # tb_presence <- tb_long %>% 
    #   dplyr::group_by(feature, dataset_name) %>% 
    #   dplyr::summarise(present = any(abundance > 0)) %>% 
    #   dplyr::group_by(feature) %>% 
    #   dplyr::mutate(n_present = sum(present),
    #                 n_absent = sum(!present))
    # tb_long <- tb_long %>% 
    #   dplyr::left_join(tb_presence,
    #                    by = c("feature", "dataset_name"))
    # tb_uniqPres <- tb_long %>% 
    #   dplyr::filter(n_present == 1, present) %>% 
    #   dplyr::group_by(feature, dataset_name) %>% 
    #   dplyr::summarise(mean_abundance = mean(abundance)) %>% 
    #   dplyr::ungroup()
    # tb_uniqAbs <- tb_long %>% 
    #   dplyr::filter(n_absent == 1, present) %>% 
    #   dplyr::group_by(feature) %>% 
    #   dplyr::summarise(mean_abundance = mean(abundance)) %>% 
    #   dplyr::ungroup() %>% 
    #   dplyr::left_join(
    #     tb_presence %>% 
    #       dplyr::filter(n_absent == 1,
    #                     !present),
    #     by = "feature"
    #   ) %>% 
    #   dplyr::select(feature, dataset_name, mean_abundance)
    # 
    # # Table summarises feature presence
    # tb_presence %>% 
    #   dplyr::filter(!duplicated(feature)) %>% 
    #   dplyr::group_by(n_present) %>% 
    #   dplyr::summarise(N = n()) %>% 
    #   t() %>% 
    #   DT::datatable(caption = feature_level) %>% 
    #   print()
    # 
    # # Figure uniquely present features
    # p <- tb_uniqPres %>% 
    #   dplyr::arrange(dataset_name, mean_abundance) %>% 
    #   dplyr::mutate(feature = factor(feature, levels = feature)) %>% 
    #   ggplot(aes(x = feature, y = mean_abundance, color = dataset_name)) + geom_point() +
    #   rotate_xaxis(90) 
    # ggsave(filename = paste0(dir_output, "fig_", feature_level, "UniquePresent.pdf"),
    #        plot = p,
    #        width = 40,
    #        height = 10)
    # p <- tb_uniqPres %>% 
    #   dplyr::filter(mean_abundance > 1e-4) %>% 
    #   dplyr::arrange(dataset_name, mean_abundance) %>% 
    #   dplyr::mutate(feature = factor(feature, levels = feature)) %>% 
    #   ggplot(aes(x = feature, y = mean_abundance, color = dataset_name)) + geom_point() +
    #   rotate_xaxis(90) + coord_flip()
    # p %>% print() %>% 
    #   ggsave(filename = paste0(dir_output, "fig_", feature_level, "UniquePresentZoom.pdf"),
    #          plot = .,
    #          width = 15,
    #          height = 10)
    # 
    # # Figure uniquely absent features
    # p <- tb_uniqAbs %>% 
    #   dplyr::arrange(dataset_name, mean_abundance) %>% 
    #   dplyr::mutate(feature = factor(feature, levels = feature)) %>% 
    #   ggplot(aes(x = feature, y = mean_abundance, color = dataset_name)) + geom_point() +
    #   rotate_xaxis(90) + coord_flip() 
    # p %>% print() %>% 
    #   ggsave(filename = paste0(dir_output, "fig_", feature_level, "UniqueAbsent.pdf"),
    #          plot = .,
    #          width = 15,
    #          height = 10)
  })
```